syntax = "proto3";
import "dtypes.proto";

package ddl_generator;

service DDLGenerator {
    rpc GenerateDDL(DDLRequest) returns (DDLResponse) {}
}

message DDLRequest {
    dtypes.AST ast = 1;
    map<string, string> columns = 2;
}

message DDLResponse {
    dtypes.AstType type = 1;
    string sql = 2;

    // Numeric, Logical, and Text Values
    oneof value {
        float number_value = 3;
        string text_value = 4;
        bool logical_value = 5;
    }

    // Cell Reference
    optional string cell = 6;
    optional dtypes.RefType refType = 7;
    optional string column = 8;
    optional string eror = 9;

    // Cell Range
    optional string start = 10;
    optional string end = 11;
    repeated string cells = 12;
    repeated string columns = 13;
    optional string error = 14;

    // Binary Expression
    optional string operator = 15;
    optional DDLResponse left = 16;
    optional DDLResponse right = 17;

    // Function
    optional string name = 18;
    repeated DDLResponse arguments = 19;

    // Unary Expression
    optional DDLResponse operand = 20;  // operator and type as well
}
